### PR TITLE
Implement get certificate count + get certificate at index

### PR DESCRIPTION
**Pull request by jchien14**
- Implements `get_certificate_count` and `get_certificate_at_index` on SecTrust to retrieve specific certificates in the chain. Unit tests added for both.
- Add the 10.12 feature, and bindings for `SecKeyCopyExternalRepresentation` and `SecItemExport`.

### CODE REVIEW

File: security-framework-sys/src/trust.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub type SecTrustRef = *mut __SecTrust;
 extern "C" {
     pub fn SecTrustGetTypeID() -> CFTypeID;
     pub fn SecTrustGetCertificateCount(trust: SecTrustRef) -> CFIndex;
-    pub fn SecTrustGetCertificateAtIndex(trust: SecTrustRef) -> SecCertificateRef;
+    pub fn SecTrustGetCertificateAtIndex(trust: SecTrustRef, ix: CFIndex) -> SecCertificateRef;
```
Comment (sfackler) [inline security-framework-sys/src/trust.rs]: Ugh, I need to get ctest set up :(

File: security-framework/src/trust.rs [hunk 2]
```diff
@@ -105,6 +106,29 @@ impl SecTrust {
             Ok(TrustResult::from_raw(result))
         }
     }
+
+    /// Returns the number of certificates in an evaluated certificate chain.
+    ///
+    /// Note: evaluate must first be called on the SecTrust.
+    pub fn get_certificate_count(&self) -> Result<i64> {
```
Comment (sfackler) [inline security-framework/src/trust.rs]: Can you remove the `get_` prefixes here and below?

File: security-framework/src/trust.rs [hunk 3]
```diff
@@ -105,6 +106,29 @@ impl SecTrust {
...
+    }
+
+    /// Returns a specific certificate from the certificate chain used to evaluate trust.
+    ///
+    /// Note: evaluate must first be called on the SecTrust.
+    pub fn certificate_at_index(&self, ix: i64) -> Result<SecCertificate> {
+        unsafe {
+            let certificate = SecTrustGetCertificateAtIndex(self.0, ix);
+            if certificate.is_null() {
+                Err(Error::from_code(errSecParam))
```
Comment (sfackler) [inline security-framework/src/trust.rs]: Seems like it'd make more sense to return an option here?

File: security-framework/src/trust.rs [hunk 4]
```diff
@@ -105,6 +106,29 @@ impl SecTrust {
             Ok(TrustResult::from_raw(result))
         }
     }
+
+    /// Returns the number of certificates in an evaluated certificate chain.
+    ///
+    /// Note: evaluate must first be called on the SecTrust.
+    pub fn certificate_count(&self) -> Result<i64> {
+        unsafe {
+            Ok(SecTrustGetCertificateCount(self.0))
```
Comment (sfackler) [inline security-framework/src/trust.rs]: Why does this return an result if it's always `Ok`?

File: security-framework/src/trust.rs [hunk 5]
```diff
@@ -105,6 +106,29 @@ impl SecTrust {
             Ok(TrustResult::from_raw(result))
         }
     }
+
+    /// Returns the number of certificates in an evaluated certificate chain.
+    ///
+    /// Note: evaluate must first be called on the SecTrust.
+    pub fn certificate_count(&self) -> i64 {
```
Comment (sfackler) [inline security-framework/src/trust.rs]: `i64` isn't going to be the right type on 32 bit, right? It probably makes sense to just return CFIndex directly.

**Comment by mako-taco** [comment]
Can we get this landed? I have to use a fork of this repo right now...

**Comment by sfackler** [comment]
The i64 use for certificate indices needs to be adjusted I believe.

File: security-framework/src/trust.rs (hunk 5 above)
Comment (cjubb39) [inline security-framework/src/trust.rs]: `isize` seems like it'd fit here?.  Also have a fork of the repo -- would like to unify :)

**Comment by jchien14** [comment]
Fixed! Although probably not a big deal for macOS since Rust only supports 10.7+ and 10.6 was the last supported 32-bit OS. Sorry for the delay

**Comment by sfackler** [comment]
Thanks!

**Pull request merged**